src/resolve/resolved-dns-browse-services.c

```diff
@@ -0,0 +1,779 @@
+                        domain = mfree(domain);
+                        r = dns_service_split(dns_resource_key_name(service->rr->key), &name, &type, &domain);
+                        if (r < 0) {
+                                log_error_errno(r,
+                                                "Failed to split DNS service name (fallback) from list: %m");
+                                goto finish;
+                        }
+                }
+
+                dns_remove_service(sb, service);
+
+                log_debug("Remove from the list %s, %s, %s, %s, %d",
+                          strna(name),
+                          strna(type),
+                          strna(domain),
+                          strna(af_to_ipv4_ipv6(owner_family)),
+                          sb->ifindex);
+
+                r = sd_json_buildo(
+                                &entry,
+                                SD_JSON_BUILD_PAIR(
+                                                "updateFlag",
+                                                SD_JSON_BUILD_STRING(browse_service_update_event_to_string(
+                                                                BROWSE_SERVICE_UPDATE_REMOVED))),
+                                SD_JSON_BUILD_PAIR("family", SD_JSON_BUILD_INTEGER(owner_family)),
+                                SD_JSON_BUILD_PAIR("name", SD_JSON_BUILD_STRING(name ?: "")),
+                                SD_JSON_BUILD_PAIR("type", SD_JSON_BUILD_STRING(type ?: "")),
+                                SD_JSON_BUILD_PAIR("domain", SD_JSON_BUILD_STRING(domain ?: "")),
+                                SD_JSON_BUILD_PAIR("ifindex", SD_JSON_BUILD_INTEGER(sb->ifindex)));
+                if (r < 0) {
+                        log_error_errno(r, "Failed to build JSON for removed service: %m");
+                        goto finish;
+                }
+
+                r = sd_json_variant_append_array(&array, entry);
+                if (r < 0) {
+                        log_error_errno(r, "Failed to append JSON entry to array: %m");
+                        goto finish;
+                }
+        }
+
+        if (!sd_json_variant_is_blank_array(array)) {
+                _cleanup_(sd_json_variant_unrefp) sd_json_variant *vm = NULL;
+
+                r = sd_json_buildo(&vm, SD_JSON_BUILD_PAIR("browserServiceData", SD_JSON_BUILD_VARIANT(array)));
+                if (r < 0) {
+                        log_error_errno(r,
+                                        "Failed to build JSON object for browser service data: %m");
+                        goto finish;
+                }
+
+                r = sd_varlink_notify(sb->link, vm);
+                if (r < 0) {
+                        log_error_errno(r, "Failed to notify via varlink: %m");
+                        goto finish;
+                }
+        }
+
+        return 0;
+
+finish:
+        log_error_errno(r, "Failed to process received services: %m");
```

Comment:
in the various cases you jump to this label you already logged an error, hence this is duplicate logging